identity-aware-proxy/IAPClient/IAPClient.cs

```diff
@@ -0,0 +1,129 @@
+                FileMode.Open, FileAccess.Read))
+            {
+                credentials = NewtonsoftJsonSerializer.Instance
+                    .Deserialize<Credentials>(fs);
+            }
+
+            // Generate a JWT signed with the service account's private key 
+            // containing a special "target_audience" claim.
+            var jwtBasedAccessToken =
+                CreateAccessToken(credentials.PrivateKey, iapClientId,
+                    credentials.ClientEmail);
+
+            var body = new Dictionary<string, string>
+            {
+                { "assertion", jwtBasedAccessToken },
+                { "grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"}
+            };
+
+            // Request an OIDC token for the Cloud IAP-secured client ID.
+            var httpClient = new HttpClient();
+            var httpContent = new FormUrlEncodedContent(body);
+            var result = httpClient.PostAsync(GoogleAuthConsts.OidcTokenUrl,
+                httpContent).Result;
+            var responseContent = result.Content.ReadAsStringAsync().Result;
+            int statusCode = (int)result.StatusCode;
+            if (statusCode < 200 || statusCode >= 300)
+            {
+                throw new HttpRequestException(string.Format("{0} {1}\n{2}",
+                    statusCode, result.ReasonPhrase, responseContent));
+            }
+            string token = JsonConvert.DeserializeObject<IapResponse>(
+                responseContent).IdToken;
+
+            // Include the OIDC token in an Authorization: Bearer header to 
+            // IAP-secured resource
+            httpClient.DefaultRequestHeaders.Authorization =
+                new AuthenticationHeaderValue("Bearer", token);
+            string response = httpClient.GetStringAsync(uri).Result;
+            return response;
+        }
+
+        static long ToUnixEpochDate(DateTime date)
+              => (long)Math.Round((date.ToUniversalTime() -
+                                   new DateTimeOffset(1970, 1, 1, 0, 0, 0,
+                                        TimeSpan.Zero)).TotalSeconds);
+
+        static string CreateAccessToken(string privateKey,
+            string iapClientId, string email)
+        {
+            var now = DateTime.UtcNow;
+            var currentTime = ToUnixEpochDate(now);
+            var expTime = ToUnixEpochDate(now.AddHours(1));
+
+            var claims = new[]
+            {
+                new Claim(JwtRegisteredClaimNames.Aud,
+                    GoogleAuthConsts.OidcTokenUrl),
+                new Claim(JwtRegisteredClaimNames.Sub, email),
+                new Claim(JwtRegisteredClaimNames.Iat, currentTime.ToString()),
+                new Claim(JwtRegisteredClaimNames.Exp, expTime.ToString()),
+                new Claim(JwtRegisteredClaimNames.Iss, email),
+
+                new Claim("target_audience", iapClientId)
+            };
+
+            SecurityKey key = new RsaSecurityKey(
```

Comment:
Can you put a comment with a link with documentation that says that encryption should use RSA256?

Reply:
I found no such docs.  I think these samples *are the docs*.

Reply:
It's [here](https://developers.google.com/identity/protocols/OAuth2ServiceAccount) Look at the <b>HTTP REST</b> part in "Preparing to make an authorized API call". I wish I'd read it when working on this sample before - perhaps it would take shorter to get it to work.

Reply:
Added link.